### PR TITLE
Improved region definition with recursive window shrinkage

### PR DESCRIPTION
**Pull request by mkanai**
I implemented a new region definition method.

Currently, regions are defined by the following steps:
1. Starting from the most significant variant, we add a flanking window (default: 1.5 Mb upstream and downstream) around the index variant to define a region
2. Remove variants in the region, and repeat the first step until none remains
3. Merge regions if they overlap.
4. If `--max-region-width` (default: 6 Mb) is supplied, any regions that exceed the size will be split into multiple regions, just by consecutively taking 6Mb regions from the beginning.

Here, the fourth step could be problematic because it does not consider association significance -- e.g., the lead variant in a large region might be around 6 Mb; and thus its LD neighbors will be split into separate regions.

In this PR, I replaced the fourth step with the following:
4. If `--max-region-width` (default: 6 Mb) is supplied, any regions that exceed the size will repeat the first step with a shrunken window size (default: current window size * 0.9) -- and continue the steps 1-4 until none exceeds the maximum size.

I think this implementation is better than 1) the current implementation because this considers association significance when splitting; and better than 2) using a shorter window size because this only splits a region that exceeds the maximum size.

Technically, this is implemented by recursively calling `generate_bed` function which is more intuitive than having a separate `split` function.

### CODE REVIEW

**Comment by mkanai** [comment]
In the R7 demo phenotypes, there are 41 regions that exceeds the 6 Mb threshold. The proposed method splits them nicely, still maintaining the minimum region length > 1.3 Mb.

File: python/make_finemap_inputs.py [hunk 1]
```diff
@@ -211,15 +203,44 @@ def generate_bed(sumstats,
...
+            bed = bed.merge().saveas()
+            bed_oversized = bed.filter(lambda x: len(x) > max_region_width).saveas()
+            print("Oversized regions")
+            print(bed_oversized)
+            if (len(bed_oversized)) > 0:
+                # keep in-size regions
+                bed1 = bed.filter(lambda x: len(x) <= max_region_width).saveas()
+                
+                # filter to oversized regions
+                sumstats = map(lambda x: filter_sumstat(x, bed_oversized), sumstats)
```
Comment (Fedja) [inline python/make_finemap_inputs.py:222]: this is unreadable to overwrite sumstats with filtered data... easy enough now but any longer function will get hard to read and error prone. you don't use filtered sumstats anywhere else but in recursion so either don't store the value but pass on the fly to recursion or then change variable name to sumstats_oversize_regions or similar

**Comment by Fedja** [comment]
yep the PR does what it advertises. 

This is an improvement compared to  the previous behavior! Off course there can be high LD variants left outside the regions (as in all regions but making them smaller increases that chance) and the LD based segmentation is really the correct way to go.  
Instead of trying to pre-segment whole genome, slightly easier is to compute LD for the index variant and choose region based on that LD.   But this should be good enough for now and these regions can be reviewed and reran as needed.

File: python/make_finemap_inputs.py (hunk 1 above)
Comment (mkanai) [reply python/make_finemap_inputs.py]: good catch! addressed

**Comment by mkanai** [comment]
Thanks Mitja! I 100% agree -- LD-based segmentation is WIP and it'd be great to have it in future releases. I will run this for all the r7 demo phenos and will share the results.

**Pull request merged**